File: go/compare.go
package pragmastat

import (
	"fmt"
)

// Metric represents the type of statistical metric being compared.
type Metric int

const (
	// MetricCenter represents the central value of a sample.
	MetricCenter Metric = iota
	// MetricSpread represents data dispersion (variability or scatter).
	MetricSpread
	// MetricShift represents the typical difference between two samples.
	MetricShift
	// MetricRatio represents the multiplicative factor between two samples.
	MetricRatio
	// MetricDisparity represents the normalized difference (effect size) between two samples.
	MetricDisparity
)

// String returns the string representation of the metric.
func (m Metric) String() string {
	switch m {
	case MetricCenter:
		return "center"
	case MetricSpread:
		return "spread"
	case MetricShift:
		return "shift"
	case MetricRatio:
		return "ratio"
	case MetricDisparity:
		return "disparity"
	default:
		return "unknown"
	}
}

// ComparisonVerdict represents the result of comparing an estimate against a threshold.
type ComparisonVerdict int

const (
	// VerdictInconclusive means not enough evidence to conclude (interval contains threshold).
	VerdictInconclusive ComparisonVerdict = iota
	// VerdictLess means the estimate is statistically less than the threshold.
	VerdictLess
	// VerdictGreater means the estimate is statistically greater than the threshold.
	VerdictGreater
)

// String returns the string representation of the verdict.
func (v ComparisonVerdict) String() string {
	switch v {
	case VerdictLess:
		return "less"
	case VerdictGreater:
		return "greater"
	case VerdictInconclusive:
		return "inconclusive"
	default:
		return "unknown"
	}
}

// Threshold represents a threshold value with a metric type and misrate for comparison.
type Threshold struct {
	Metric  Metric
	Value   Measurement
	Misrate float64
}

// NewThreshold creates a new Threshold with validation.
// The misrate must be in (0, 1] and the value must be finite.
func NewThreshold(metric Metric, value Measurement, misrate float64) (*Threshold, error) {
	if !misrateIsValid(misrate) {
		return nil, NewDomainError(SubjectMisrate)
	}
	if !isFinite(value.Value) {
		return nil, fmt.Errorf("threshold value must be finite")
	}
	return &Threshold{Metric: metric, Value: value, Misrate: misrate}, nil
}

// Projection represents the result of comparing an estimate against a threshold.
type Projection struct {
	Threshold Threshold
	Estimate  Measurement
	Bounds    Bounds
	Verdict   ComparisonVerdict
}

// metricSpec defines the specification for a metric's validation, estimation, and bounds computation.
type metricSpec struct {
	metric               Metric
	validateAndNormalize func(threshold *Threshold, x, y *Sample) (Measurement, error)
	estimate             func(x, y *Sample) (Measurement, error)
	bounds               func(x, y *Sample, misrate float64) (Bounds, error)
	seededBounds         func(x, y *Sample, misrate float64, seed string) (Bounds, error)
}

// compare1Specs defines the metric specifications for Compare1.
var compare1Specs = []metricSpec{
	{
		metric: MetricCenter,
		validateAndNormalize: func(threshold *Threshold, x, y *Sample) (Measurement, error) {
			return validateCenterOrSpread(threshold, x)
		},
		estimate: func(x, y *Sample) (Measurement, error) {
			return Center(x)
		},
		bounds: func(x, y *Sample, misrate float64) (Bounds, error) {
			return CenterBounds(x, misrate)
		},
	},
	{
		metric: MetricSpread,
		validateAndNormalize: func(threshold *Threshold, x, y *Sample) (Measurement, error) {
			return validateCenterOrSpread(threshold, x)
		},
		estimate: func(x, y *Sample) (Measurement, error) {
			return Spread(x)
		},
		bounds: func(x, y *Sample, misrate float64) (Bounds, error) {
			return SpreadBounds(x, misrate)
		},
		seededBounds: func(x, y *Sample, misrate float64, seed string) (Bounds, error) {
			return SpreadBoundsWithSeed(x, misrate, seed)
		},
	},
}

// compare2Specs defines the metric specifications for Compare2.
var compare2Specs = []metricSpec{
	{
		metric:               MetricShift,
		validateAndNormalize: validateShift,
		estimate:             Shift,
		bounds:               ShiftBounds,
	},
	{
		metric: MetricRatio,
		validateAndNormalize: func(threshold *Threshold, x, y *Sample) (Measurement, error) {
			return validateRatio(threshold)
		},
		estimate: Ratio,
		bounds:   RatioBounds,
	},
	{
		metric: MetricDisparity,
		validateAndNormalize: func(threshold *Threshold, x, y *Sample) (Measurement, error) {
			return validateDisparity(threshold)
		},
		estimate:     Disparity,
		bounds:       DisparityBounds,
		seededBounds: DisparityBoundsWithSeed,
	},
}

// validateCenterOrSpread validates and normalizes a Center or Spread threshold.
func validateCenterOrSpread(threshold *Threshold, x *Sample) (Measurement, error) {
	if !threshold.Value.Unit.IsCompatible(x.Unit) {
		return Measurement{}, &UnitMismatchError{Unit1: threshold.Value.Unit, Unit2: x.Unit}
	}
	factor := ConversionFactor(threshold.Value.Unit, x.Unit)
	return NewMeasurement(threshold.Value.Value*factor, x.Unit), nil
}

// validateShift validates and normalizes a Shift threshold.
func validateShift(threshold *Threshold, x, y *Sample) (Measurement, error) {
	if !threshold.Value.Unit.IsCompatible(x.Unit) {
		return Measurement{}, &UnitMismatchError{Unit1: threshold.Value.Unit, Unit2: x.Unit}
	}
	target := Finer(x.Unit, y.Unit)
	factor := ConversionFactor(threshold.Value.Unit, target)
	return NewMeasurement(threshold.Value.Value*factor, target), nil
}

// validateRatio validates and normalizes a Ratio threshold.
func validateRatio(threshold *Threshold) (Measurement, error) {
	unit := threshold.Value.Unit
	if unit != RatioUnit && unit != NumberUnit {
		return Measurement{}, &UnitMismatchError{Unit1: unit, Unit2: RatioUnit}
	}
	value := threshold.Value.Value
	if value <= 0 || !isFinite(value) {
		return Measurement{}, fmt.Errorf("ratio threshold value must be finite and positive")
	}
	return NewMeasurement(value, RatioUnit), nil
}

// validateDisparity validates and normalizes a Disparity threshold.
func validateDisparity(threshold *Threshold) (Measurement, error) {
	unit := threshold.Value.Unit
	if unit != DisparityUnit && unit != NumberUnit {
		return Measurement{}, &UnitMismatchError{Unit1: unit, Unit2: DisparityUnit}
	}
	value := threshold.Value.Value
	if !isFinite(value) {
		return Measurement{}, fmt.Errorf("disparity threshold value must be finite")
	}
	return NewMeasurement(value, DisparityUnit), nil
}

// misrateIsValid checks if the misrate is valid (in (0, 1]).
func misrateIsValid(misrate float64) bool {
	return isFinite(misrate) && misrate > 0 && misrate <= 1
}

// computeVerdict computes the verdict by comparing bounds against a threshold value.
func computeVerdict(bounds Bounds, thresholdValue float64) ComparisonVerdict {
	if bounds.Lower > thresholdValue {
		return VerdictGreater
	}
	if bounds.Upper < thresholdValue {
		return VerdictLess
	}
	return VerdictInconclusive
}

// getSpec returns the metric spec for the given metric from the provided specs.
func getSpec(specs []metricSpec, metric Metric) (*metricSpec, error) {
	for i := range specs {
		if specs[i].metric == metric {
			return &specs[i], nil
		}
	}
	return nil, fmt.Errorf("no spec found for metric %v", metric)
}

// Compare1 performs one-sample confirmatory analysis: compares Center/Spread against practical thresholds.
//
// Parameters:
//   - x: The sample to analyze
//   - thresholds: List of thresholds to compare against (must not be empty)
//
// Returns a slice of Projections in the same order as the input thresholds.
func Compare1(x *Sample, thresholds []*Threshold) ([]Projection, error) {
	return Compare1WithSeed(x, thresholds, "")
}

// Compare1WithSeed performs one-sample confirmatory analysis with a seed for reproducibility.
// The seed is used for randomized bounds (Spread bounds only).
func Compare1WithSeed(x *Sample, thresholds []*Threshold, seed string) ([]Projection, error) {
	if err := checkNonWeighted("x", x); err != nil {
		return nil, err
	}

	if len(thresholds) == 0 {
		return nil, fmt.Errorf("thresholds list cannot be empty")
	}

	// Validate all thresholds are supported by Compare1
	for i, threshold := range thresholds {
		if threshold == nil {
			return nil, fmt.Errorf("thresholds[%d] cannot be nil", i)
		}
		if threshold.Metric != MetricCenter && threshold.Metric != MetricSpread {
			return nil, fmt.Errorf("metric %v is not supported by Compare1. Use Compare2 instead", threshold.Metric)
		}
	}

	// Normalize all threshold values
	normalizedValues := make([]Measurement, len(thresholds))
	for i, threshold := range thresholds {
		spec, err := getSpec(compare1Specs, threshold.Metric)
		if err != nil {
			return nil, err
		}
		normalized, err := spec.validateAndNormalize(threshold, x, nil)
		if err != nil {
			return nil, err
		}
		normalizedValues[i] = normalized
	}

	return executeCompare(compare1Specs, x, nil, thresholds, normalizedValues, seed)
}

// Compare2 performs two-sample confirmatory analysis: compares Shift/Ratio/Disparity against practical thresholds.
//
// Parameters:
//   - x: The first sample
//   - y: The second sample
//   - thresholds: List of thresholds to compare against (must not be empty)
//
// Returns a slice of Projections in the same order as the input thresholds.
func Compare2(x, y *Sample, thresholds []*Threshold) ([]Projection, error) {
	return Compare2WithSeed(x, y, thresholds, "")
}

// Compare2WithSeed performs two-sample confirmatory analysis with a seed for reproducibility.
// The seed is used for randomized bounds (Disparity bounds only).
func Compare2WithSeed(x, y *Sample, thresholds []*Threshold, seed string) ([]Projection, error) {
	if err := checkNonWeighted("x", x); err != nil {
		return nil, err
	}
	if err := checkNonWeighted("y", y); err != nil {
		return nil, err
	}
	if err := checkCompatibleUnits(x, y); err != nil {
		return nil, err
	}

	if len(thresholds) == 0 {
		return nil, fmt.Errorf("thresholds list cannot be empty")
	}

	// Validate all thresholds are supported by Compare2
	for i, threshold := range thresholds {
		if threshold == nil {
			return nil, fmt.Errorf("thresholds[%d] cannot be nil", i)
		}
		if threshold.Metric != MetricShift && threshold.Metric != MetricRatio && threshold.Metric != MetricDisparity {
			return nil, fmt.Errorf("metric %v is not supported by Compare2. Use Compare1 instead", threshold.Metric)
		}
	}

	// Convert both samples to the finer unit before any estimation
	target := Finer(x.Unit, y.Unit)
	xConv, err := x.ConvertTo(target)
	if err != nil {
		return nil, err
	}
	yConv, err := y.ConvertTo(target)
	if err != nil {
		return nil, err
	}

	// Normalize all threshold values
	normalizedValues := make([]Measurement, len(thresholds))
	for i, threshold := range thresholds {
		spec, err := getSpec(compare2Specs, threshold.Metric)
		if err != nil {
			return nil, err
		}
		normalized, err := spec.validateAndNormalize(threshold, xConv, yConv)
		if err != nil {
			return nil, err
		}
		normalizedValues[i] = normalized
	}

	return executeCompare(compare2Specs, xConv, yConv, thresholds, normalizedValues, seed)
}

// executeCompare executes the comparison for the given specs and returns projections.
func executeCompare(
	specs []metricSpec,
	x, y *Sample,
	thresholds []*Threshold,
	normalizedValues []Measurement,
	seed string,
) ([]Projection, error) {
	results := make([]Projection, len(thresholds))
	computed := make([]bool, len(thresholds))

	// Group thresholds by metric
	type entry struct {
		index           int
		threshold       *Threshold
		normalizedValue Measurement
	}
	byMetric := make(map[Metric][]entry)

	for i, threshold := range thresholds {
		byMetric[threshold.Metric] = append(byMetric[threshold.Metric], entry{
			index:           i,
			threshold:       threshold,
			normalizedValue: normalizedValues[i],
		})
	}

	// Process each metric spec
	for i := range specs {
		spec := &specs[i]
		entries, ok := byMetric[spec.metric]
		if !ok {
			continue
		}

		// Compute estimate once per metric
		estimate, err := spec.estimate(x, y)
		if err != nil {
			return nil, err
		}

		// Compute bounds and verdict for each threshold of this metric
		for _, e := range entries {
			var bounds Bounds
			if seed != "" && spec.seededBounds != nil {
				bounds, err = spec.seededBounds(x, y, e.threshold.Misrate, seed)
			} else {
				bounds, err = spec.bounds(x, y, e.threshold.Misrate)
			}
			if err != nil {
				return nil, err
			}

			verdict := computeVerdict(bounds, e.normalizedValue.Value)
			results[e.index] = Projection{
				Threshold: *e.threshold,
				Estimate:  estimate,
				Bounds:    bounds,
				Verdict:   verdict,
			}
			computed[e.index] = true
		}
	}

	// Verify all projections were computed
	for i, ok := range computed {
		if !ok {
			return nil, fmt.Errorf("projection %d was not computed", i)
		}
	}

	return results, nil
}

File: ts/src/compare.ts
/**
 * Compare1 and Compare2: confirmatory analysis for one-sample and two-sample estimators.
 *
 * These high-level APIs compare estimates (Center, Spread, Shift, Ratio, Disparity)
 * against practical thresholds and return verdicts (Less, Greater, or Inconclusive).
 */

import { Measurement } from './measurement';
import { MeasurementUnit } from './measurement-unit';
import { Sample, checkNonWeighted, checkCompatibleUnits, convertToFiner } from './sample';
import { AssumptionError } from './assumptions';
import {
  center,
  spread,
  shift,
  ratio,
  disparity,
  centerBounds,
  spreadBounds,
  shiftBounds,
  ratioBounds,
  disparityBounds,
  Bounds,
} from './estimators';

/**
 * Metric types supported by Compare1 and Compare2.
 */
export enum Metric {
  Center = 'center',
  Spread = 'spread',
  Shift = 'shift',
  Ratio = 'ratio',
  Disparity = 'disparity',
}

/**
 * Verdict from comparing an estimate against a threshold.
 */
export enum ComparisonVerdict {
  Less = 'less',
  Greater = 'greater',
  Inconclusive = 'inconclusive',
}

/**
 * A threshold value with a metric type and misrate for comparison.
 */
export class Threshold {
  constructor(
    readonly metric: Metric,
    readonly value: Measurement,
    readonly misrate: number,
  ) {
    if (!Number.isFinite(misrate) || misrate <= 0.0 || misrate > 1.0) {
      throw AssumptionError.domain('misrate');
    }
    if (!Number.isFinite(value.value)) {
      throw new Error('threshold value must be finite');
    }
  }
}

/**
 * A projection containing estimate, bounds, and verdict for a single threshold.
 */
export class Projection {
  constructor(
    readonly threshold: Threshold,
    readonly estimate: Measurement,
    readonly bounds: Bounds,
    readonly verdict: ComparisonVerdict,
  ) {}
}

/** Function type for validating and normalizing threshold values. */
type ValidateAndNormalizeFn = (threshold: Threshold, x: Sample, y: Sample | null) => Measurement;

/** Function type for computing point estimates. */
type EstimateFn = (x: Sample, y: Sample | null) => Measurement;

/** Function type for computing bounds. */
type BoundsFn = (x: Sample, y: Sample | null, misrate: number, seed?: string) => Bounds;

/** Specification for a metric's validation, estimation, and bounds computation. */
interface MetricSpec {
  metric: Metric;
  validateAndNormalize: ValidateAndNormalizeFn;
  estimate: EstimateFn;
  bounds: BoundsFn;
  seededBounds?: BoundsFn;
}

/** Specification for Compare2 metrics (validation only — estimation is done directly). */
interface Compare2MetricSpec {
  metric: Metric;
  validateAndNormalize: ValidateAndNormalizeFn;
}

/** Validates and normalizes a Center threshold. */
function validateCenter(threshold: Threshold, x: Sample, _y: Sample | null): Measurement {
  if (!threshold.value.unit.isCompatible(x.unit)) {
    throw new AssumptionError(
      `can't convert ${threshold.value.unit.fullName} to ${x.unit.fullName}`,
    );
  }
  const factor = MeasurementUnit.conversionFactor(threshold.value.unit, x.unit);
  return new Measurement(threshold.value.value * factor, x.unit);
}

/** Validates and normalizes a Spread threshold. */
function validateSpread(threshold: Threshold, x: Sample, _y: Sample | null): Measurement {
  return validateCenter(threshold, x, null); // Same validation as Center
}

/** Validates and normalizes a Shift threshold. */
function validateShift(threshold: Threshold, x: Sample, y: Sample | null): Measurement {
  if (!y) {
    throw new Error('Shift requires y sample');
  }
  if (!threshold.value.unit.isCompatible(x.unit)) {
    throw new AssumptionError(
      `can't convert ${threshold.value.unit.fullName} to ${x.unit.fullName}`,
    );
  }
  const target = MeasurementUnit.finer(x.unit, y.unit);
  const factor = MeasurementUnit.conversionFactor(threshold.value.unit, target);
  return new Measurement(threshold.value.value * factor, target);
}

/** Validates and normalizes a Ratio threshold. */
function validateRatio(threshold: Threshold, _x: Sample, _y: Sample | null): Measurement {
  const unit = threshold.value.unit;
  if (unit.id !== 'ratio' && unit.id !== 'number') {
    throw new AssumptionError(`can't convert ${unit.fullName} to Ratio`);
  }
  const value = threshold.value.value;
  if (value <= 0.0) {
    throw new Error('Ratio threshold value must be positive');
  }
  return new Measurement(value, MeasurementUnit.RATIO);
}

/** Validates and normalizes a Disparity threshold. */
function validateDisparity(threshold: Threshold, _x: Sample, _y: Sample | null): Measurement {
  const unit = threshold.value.unit;
  if (unit.id !== 'disparity' && unit.id !== 'number') {
    throw new AssumptionError(`can't convert ${unit.fullName} to Disparity`);
  }
  return new Measurement(threshold.value.value, MeasurementUnit.DISPARITY);
}

/** Compare1 metric specifications. */
const compare1Specs: MetricSpec[] = [
  {
    metric: Metric.Center,
    validateAndNormalize: validateCenter,
    estimate: (x, _y) => center(x),
    bounds: (x, _y, misrate, _seed?) => centerBounds(x, misrate),
  },
  {
    metric: Metric.Spread,
    validateAndNormalize: validateSpread,
    estimate: (x, _y) => spread(x),
    bounds: (x, _y, misrate, seed?) => spreadBounds(x, misrate, seed),
    seededBounds: (x, _y, misrate, seed?) => spreadBounds(x, misrate, seed),
  },
];

/** Compare2 metric specifications. */
const compare2Specs: Compare2MetricSpec[] = [
  { metric: Metric.Shift, validateAndNormalize: validateShift },
  { metric: Metric.Ratio, validateAndNormalize: validateRatio },
  { metric: Metric.Disparity, validateAndNormalize: validateDisparity },
];

/** Computes the verdict by comparing bounds against a threshold value. */
function computeVerdict(bounds: Bounds, thresholdValue: number): ComparisonVerdict {
  if (bounds.lower > thresholdValue) {
    return ComparisonVerdict.Greater;
  }
  if (bounds.upper < thresholdValue) {
    return ComparisonVerdict.Less;
  }
  return ComparisonVerdict.Inconclusive;
}

/** Gets the spec for a given metric from the spec array. */
function getSpec<T extends { metric: Metric }>(specs: T[], metric: Metric): T | undefined {
  return specs.find((s) => s.metric === metric);
}

/** Executes Compare1 logic. */
function executeCompare1(
  x: Sample,
  thresholds: Threshold[],
  normalizedValues: Measurement[],
  seed?: string,
): Projection[] {
  const results: (Projection | null)[] = new Array(thresholds.length).fill(null);

  // Group thresholds by metric
  const byMetric = new Map<
    Metric,
    Array<{ index: number; threshold: Threshold; normalizedValue: Measurement }>
  >();

  for (let i = 0; i < thresholds.length; i++) {
    const metric = thresholds[i].metric;
    if (!byMetric.has(metric)) {
      byMetric.set(metric, []);
    }
    byMetric.get(metric)!.push({
      index: i,
      threshold: thresholds[i],
      normalizedValue: normalizedValues[i],
    });
  }

  // Process each metric spec
  for (const spec of compare1Specs) {
    const entries = byMetric.get(spec.metric);
    if (!entries || entries.length === 0) {
      continue;
    }

    // Compute estimate once per metric
    const estimate = spec.estimate(x, null);

    // Compute bounds and verdict for each threshold of this metric
    for (const entry of entries) {
      const bounds =
        seed && spec.seededBounds
          ? spec.seededBounds(x, null, entry.threshold.misrate, seed)
          : spec.bounds(x, null, entry.threshold.misrate);

      const verdict = computeVerdict(bounds, entry.normalizedValue.value);
      results[entry.index] = new Projection(entry.threshold, estimate, bounds, verdict);
    }
  }

  return results.filter((r): r is Projection => r !== null);
}

/** Executes Compare2 logic. */
function executeCompare2(
  x: Sample,
  y: Sample,
  thresholds: Threshold[],
  normalizedValues: Measurement[],
  seed?: string,
): Projection[] {
  const results: (Projection | null)[] = new Array(thresholds.length).fill(null);

  // Convert both samples to the finer unit
  const [cx, cy] = convertToFiner(x, y);

  // Group thresholds by metric
  const byMetric = new Map<
    Metric,
    Array<{ index: number; threshold: Threshold; normalizedValue: Measurement }>
  >();

  for (let i = 0; i < thresholds.length; i++) {
    const metric = thresholds[i].metric;
    if (!byMetric.has(metric)) {
      byMetric.set(metric, []);
    }
    byMetric.get(metric)!.push({
      index: i,
      threshold: thresholds[i],
      normalizedValue: normalizedValues[i],
    });
  }

  // Process Shift thresholds
  const shiftEntries = byMetric.get(Metric.Shift);
  if (shiftEntries && shiftEntries.length > 0) {
    const estimate = shift(cx, cy);
    for (const entry of shiftEntries) {
      const bounds = shiftBounds(cx, cy, entry.threshold.misrate);
      const verdict = computeVerdict(bounds, entry.normalizedValue.value);
      results[entry.index] = new Projection(entry.threshold, estimate, bounds, verdict);
    }
  }

  // Process Ratio thresholds
  const ratioEntries = byMetric.get(Metric.Ratio);
  if (ratioEntries && ratioEntries.length > 0) {
    const estimate = ratio(cx, cy);
    for (const entry of ratioEntries) {
      const bounds = ratioBounds(cx, cy, entry.threshold.misrate);
      const verdict = computeVerdict(bounds, entry.normalizedValue.value);
      results[entry.index] = new Projection(entry.threshold, estimate, bounds, verdict);
    }
  }

  // Process Disparity thresholds
  const disparityEntries = byMetric.get(Metric.Disparity);
  if (disparityEntries && disparityEntries.length > 0) {
    const estimate = disparity(cx, cy);
    for (const entry of disparityEntries) {
      const bounds = disparityBounds(cx, cy, entry.threshold.misrate, seed);
      const verdict = computeVerdict(bounds, entry.normalizedValue.value);
      results[entry.index] = new Projection(entry.threshold, estimate, bounds, verdict);
    }
  }

  return results.filter((r): r is Projection => r !== null);
}

/**
 * One-sample confirmatory analysis: compares Center/Spread against practical thresholds.
 *
 * @param x The sample to analyze
 * @param thresholds List of thresholds to compare against
 * @returns List of projections in the same order as the input thresholds
 * @throws Error if thresholds is empty or contains unsupported metrics
 * @throws AssumptionError if validation fails
 */
export function compare1(x: Sample, thresholds: Threshold[]): Projection[];
export function compare1(x: Sample, thresholds: Threshold[], seed: string): Projection[];
export function compare1(x: Sample, thresholds: Threshold[], seed?: string): Projection[] {
  checkNonWeighted('x', x);

  if (thresholds.length === 0) {
    throw new Error('thresholds list cannot be empty');
  }

  for (const threshold of thresholds) {
    if (threshold.metric !== Metric.Center && threshold.metric !== Metric.Spread) {
      throw new Error(
        `Metric ${threshold.metric} is not supported by Compare1. Use Compare2 instead.`,
      );
    }
  }

  const normalizedValues: Measurement[] = [];
  for (const threshold of thresholds) {
    const spec = getSpec(compare1Specs, threshold.metric);
    if (!spec) {
      throw new Error(`No spec found for metric ${threshold.metric}`);
    }
    const normalized = spec.validateAndNormalize(threshold, x, null);
    normalizedValues.push(normalized);
  }

  return executeCompare1(x, thresholds, normalizedValues, seed);
}

/**
 * Two-sample confirmatory analysis: compares Shift/Ratio/Disparity against practical thresholds.
 *
 * @param x The first sample
 * @param y The second sample
 * @param thresholds List of thresholds to compare against
 * @returns List of projections in the same order as the input thresholds
 * @throws Error if thresholds is empty or contains unsupported metrics
 * @throws AssumptionError if validation fails
 */
export function compare2(x: Sample, y: Sample, thresholds: Threshold[]): Projection[];
export function compare2(x: Sample, y: Sample, thresholds: Threshold[], seed: string): Projection[];
export function compare2(
  x: Sample,
  y: Sample,
  thresholds: Threshold[],
  seed?: string,
): Projection[] {
  checkNonWeighted('x', x);
  checkNonWeighted('y', y);
  checkCompatibleUnits(x, y);

  if (thresholds.length === 0) {
    throw new Error('thresholds list cannot be empty');
  }

  for (const threshold of thresholds) {
    if (
      threshold.metric !== Metric.Shift &&
      threshold.metric !== Metric.Ratio &&
      threshold.metric !== Metric.Disparity
    ) {
      throw new Error(
        `Metric ${threshold.metric} is not supported by Compare2. Use Compare1 instead.`,
      );
    }
  }

  // Normalize threshold values
  const normalizedValues: Measurement[] = [];
  for (const threshold of thresholds) {
    const spec = getSpec(compare2Specs, threshold.metric);
    if (!spec) {
      throw new Error(`No spec found for metric ${threshold.metric}`);
    }
    const normalized = spec.validateAndNormalize(threshold, x, y);
    normalizedValues.push(normalized);
  }

  return executeCompare2(x, y, thresholds, normalizedValues, seed);
}
